### ISSUE TITLE
randomwalker error

### ISSUE THREAD

**Issue by Phlya** [scikit-image/scikit-image#2783]
## Description
I am trying to use the randomwalker segmentation method, which seems to be generally superior to watershed. However on some images or with certain pre-processing I get a strange error. When I apply the same code to the same image with different gaussian smoothing (i.e. different sigma) it can either work, or not! The same applies to identically processed z-slices of the same 3D-image, for example.
Here is a simplified version of my code:
```
def apply_randomwalker2D(image):
    distance = ndimage.distance_transform_edt(image)
    distance = np.array(distance)
    local_maxi = peak_local_max(distance, min_distance=10, num_peaks=2,
                                indices=False, labels=image)
    markers = label(local_maxi)
    markers[image==0] = -1
    labels_rw = skimage.segmentation.random_walker(image, markers, mode='bf')
    labels_rw[labels_rw==-1]=0
    return labels_rw

def segment_z(subim):
    a = gaussian(subim, 3)
    mask = a>threshold_otsu(a)
    regions = apply_randomwalker2D(ndimage.binary_fill_holes(mask))
    regions[~mask]=0
    regions = skimage.segmentation.relabel_sequential(regions)[0]
    return regions
```

Here is the traceback:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-388-aafa10950978> in <module>()
      1 plt.imshow(im[61], cmap='Greys_r', vmax=im.max())
----> 2 regions = segment_z(im[61])
      3 plt.imshow(np.ma.masked_equal(regions, 0), vmin=0, vmax=regions.max(), cmap='viridis', alpha=0.15)
      4 plt.show()

<ipython-input-387-c57927be1e22> in segment_z(subim)
     17 
     18     mask = a>threshold_otsu(a)
---> 19     regions = apply_randomwalker2D(ndimage.binary_fill_holes(mask))
     20     regions[~mask]=0
     21     regions = skimage.segmentation.relabel_sequential(regions)[0]

<ipython-input-387-c57927be1e22> in apply_randomwalker2D(image)
      6     markers = label(local_maxi)
      7     markers[image==0] = -1
----> 8     labels_rw = skimage.segmentation.random_walker(image, markers, mode='bf')
      9     labels_rw[labels_rw==-1]=0
     10     return labels_rw

/usr/local/lib/python3.5/dist-packages/skimage/segmentation/random_walker_segmentation.py in random_walker(data, labels, beta, mode, tol, copy, multichannel, return_full_prob, spacing)
    433         lap_sparse = _build_laplacian(data, spacing, beta=beta,
    434                                       multichannel=multichannel)
--> 435     lap_sparse, B = _buildAB(lap_sparse, labels)
    436 
    437     # We solve the linear system

/usr/local/lib/python3.5/dist-packages/skimage/segmentation/random_walker_segmentation.py in _buildAB(lap_sparse, labels)
    145     seeds_indices = indices[labels > 0]
    146     # The following two lines take most of the time in this function
--> 147     B = lap_sparse[unlabeled_indices][:, seeds_indices]
    148     lap_sparse = lap_sparse[unlabeled_indices][:, unlabeled_indices]
    149     nlabels = labels.max()

/usr/local/lib/python3.5/dist-packages/scipy/sparse/csr.py in __getitem__(self, key)
    314             # [[1,2],??]
    315             if isintlike(col) or isinstance(col,slice):
--> 316                 P = extractor(row, self.shape[0])     # [[1,2],j] or [[1,2],1:2]
    317                 extracted = P * self
    318                 if col == slice(None, None, None):

/usr/local/lib/python3.5/dist-packages/scipy/sparse/csr.py in extractor(indices, N)
    268             indices = asindices(indices)
    269 
--> 270             (min_indx,max_indx) = check_bounds(indices,N)
    271 
    272             if min_indx < 0:

/usr/local/lib/python3.5/dist-packages/scipy/sparse/csr.py in check_bounds(indices, N)
    254             max_indx = indices.max()
    255             if max_indx >= N:
--> 256                 raise IndexError('index (%d) out of range' % max_indx)
    257 
    258             min_indx = indices.min()

IndexError: index (64395) out of range
```
If I change the sigma to, for example, 5, it works fine! Here is the txt file with the array of this image to test:
[array.txt](https://github.com/scikit-image/scikit-image/files/1260959/array.txt)

This is what the segmentation looks like with sigma=5.
![index](https://user-images.githubusercontent.com/2895034/29831956-9ff67d2e-8cde-11e7-84f5-d640e4d536e6.png)

I am using skimage 0.13.0. Let me know if versions of any other packages are important...

**Comment by emmanuelle** [comment]
This looks like a bug. I'm looking into it.

**Comment by emmanuelle** [comment]
OK, so here's a minimal example reproducing the bug. The problem appears when you have seed pixels which are not connected to unlabeled pixels, but just to background pixels. When you change the value of sigma unlabeled regions get typically bigger to you don't have this problem. 

```
import numpy as np
from skimage import segmentation

a = np.zeros((7, 7))
mask = - np.ones(a.shape)
mask[1, 1] = 1
mask[4:, 4:] = 0 
mask[5, 5] = 2

res = segmentation.random_walker(a, mask)
```

**Comment by emmanuelle** [comment]
As a quick workaround I left out isolated seed regions, as in the code below. You can use this solution until we correct the bug.
```
from scipy import ndimage
import skimage
from skimage import filters, feature, measure
import numpy as np
import matplotlib.pyplot as plt

def apply_randomwalker2D(image):
    distance = ndimage.distance_transform_edt(image)
    distance = np.array(distance)
    local_maxi = feature.peak_local_max(distance, min_distance=10, num_peaks=2,
                                indices=False, labels=image)
    markers = measure.label(local_maxi)
    markers[image==0] = -1
    fill = ndimage.binary_propagation(markers == 0, mask=markers>=0)
    isolated = np.logical_and(markers>0, np.logical_not(fill))
    markers_new = np.copy(markers)
    # Ignore isolated seeds
    markers_new[isolated] = -1
    labels_rw = skimage.segmentation.random_walker(image, markers_new,
                                                   mode='bf')
    labels_rw[labels_rw==-1]=0
    # Put back labels of isolated seeds.
    labels_rw[isolated] = markers[isolated]
    return labels_rw

def segment_z(subim):
    a = filters.gaussian(subim, 3)
    mask = a > filters.threshold_otsu(a)
    regions = apply_randomwalker2D(ndimage.binary_fill_holes(mask))
    regions[~mask]=0
    regions = skimage.segmentation.relabel_sequential(regions)[0]
    return regions

a = np.loadtxt('array.txt')
regions = segment_z(a)
plt.imshow(regions)
plt.show()
```

**Comment by Phlya** [comment]
Thanks for looking into it and suggesting a workaround @emmanuelle!

**Comment by emmanuelle** [comment]
You're welcome @Phlya Sorry for the inconvenience with the bug!

**Comment by stefanv** [comment]
@emmanuelle Any luck in figuring out a patch for this problem?

**Comment by emmanuelle** [comment]
@stefanv not yet, hope to work on it this week-end

**Comment by jni** [comment]
Closed by #2946
